tools/wal-reader/main.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/prometheus/prometheus/tsdb/record"
+	"github.com/prometheus/prometheus/tsdb/wlog"
+
+	util_math "github.com/grafana/mimir/pkg/util/math"
+)
+
+func main() {
+	// Clean up all flags registered via init() methods of 3rd-party libraries.
+	flag.CommandLine = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	var (
+		printSeriesEntries           bool
+		printSeriesWithSampleEntries bool
+	)
+
+	flag.BoolVar(&printSeriesEntries, "series-entries", true, "Print series entries")
+	flag.BoolVar(&printSeriesWithSampleEntries, "print-series-with-samples", false, "Print series information for each sample")
+	args, err := flagext.ParseFlagsAndArguments(flag.CommandLine)
+	if err != nil {
+		log.Fatalln(err.Error())
+	}
+
+	log.SetOutput(os.Stdout)
+
+	for _, dir := range args {
+		err := printWal(dir, printSeriesEntries, printSeriesWithSampleEntries)
+		if err != nil {
+			log.Fatalln("failed to print WAL from directory", dir, "due to error:", err)
+		}
+	}
+}
+
+func printWal(walDir string, printSeriesEntries, printSeriesWithSampleEntries bool) error {
+	var minSampleTime, maxSampleTime int64 = math.MaxInt64, math.MinInt64
+	// map series refs to labels (as a string)
+	series := map[chunks.HeadSeriesRef]string{}
+
+	// Backfill the checkpoint first if it exists.
+	checkpointDir, startFrom, err := wlog.LastCheckpoint(walDir)
+	if err != nil && !errors.Is(err, record.ErrNotFound) {
+		return errors.Wrap(err, "find last checkpoint")
+	}
+
+	if err == nil {
+		log.Println("Using checkpoint directory:", checkpointDir)
+
+		sr, err := wlog.NewSegmentsReader(checkpointDir)
+		if err != nil {
+			return errors.Wrap(err, "open checkpoint")
+		}
+		defer func() {
+			if err := sr.Close(); err != nil {
+				log.Println("Error while closing the wal segments reader when processing checkpoint", "err", err)
+			}
+		}()
+
+		log.Println("replaying checkpoint")
+		if err := printWalEntries(wlog.NewReader(sr), series, printSeriesEntries, printSeriesWithSampleEntries, &minSampleTime, &maxSampleTime); err != nil {
+			return errors.Wrap(err, "replaying checkpoint")
+		}
+		log.Println("checking replay finished")
+	} else {
+		log.Println("no checkpoint found, skipping")
+	}
+
+	// Find the last segment.
+	_, endAt, e := wlog.Segments(walDir)
+	if e != nil {
+		return errors.Wrap(e, "finding WAL segments")
+	}
+
+	startFrom++
+	for i := startFrom; i <= endAt; i++ {
+		log.Println("replaying WAL segment", i)
+		s, err := wlog.OpenReadSegment(wlog.SegmentName(walDir, i))
+		if err != nil {
+			return errors.Wrapf(err, "open WAL segment: %d", i)
+		}
+
+		sr, err := wlog.NewSegmentBufReaderWithOffset(0, s)
+		if errors.Is(err, io.EOF) {
+			// File does not exist.
+			continue
+		}
+		if err != nil {
+			return errors.Wrapf(err, "reader for segment %d", i)
+		}
+		err = printWalEntries(wlog.NewReader(sr), series, printSeriesEntries, printSeriesWithSampleEntries, &minSampleTime, &maxSampleTime)
+		if err := sr.Close(); err != nil {
```

Comment:
[nit] I would rename `err` to `closeErr` just for clarity. The different scope makes it a bit tricky to read this code.

Reply:
Done in https://github.com/grafana/mimir/pull/5780/commits/c879a353c126fc2d550761cf3ddb2864d583667d.